Specify a specific SAPUI5 Version

mathiasmaerker opened the pull request:
Hi, 
I created an enhancement for your module (which by the way I really like :+1:  )
In our project we have sometimes the need for a specific SAPUI5 version, because the update cycles on Netweaver Systems is not as high as in the cloud.

I implemented the possibility to install a specific version, defined in the projects package.json ( please see readme.md)

This PR introduces no additional dependencies nor is it a breaking change
I tagged the commit with v1.2.0 and updated the package.json accordingly

I hope you like the changes :smile: 

Best regards 
Mathias

mathiasmaerker committed:
feat(versions): Added support for specific runtime version

mathiasmaerker committed:
feat(versions): Added support for specific runtime version

mathiasmaerker committed:
Merge branch 'master' of https://github.com/mathiasmaerker/sapui5-runtime

mathiasmaerker committed:
fix(versions)

mathiasmaerker committed:
Merge branch 'master' of https://github.com/mathiasmaerker/sapui5-runtime

mathiasmaerker committed:
Merge branch 'master' of https://github.com/mathiasmaerker/sapui5-runtime

mathiasmaerker committed:
Merge branch 'master' of https://github.com/mathiasmaerker/sapui5-runtime

mathiasmaerker committed:
feat(versions): Finalized version feature

mathiasmaerker committed:
fix(versions): eslint errors

bastilimbach commented:
Hi @mathiasmaerker, thank you for your contribution. This is awesome! This feature was already on my roadmap. I'll take a look at your implementation this evening. 👍

mathiasmaerker committed:
fix(versions): fixed ENOENT error

bastilimbach commented on README.md:
```diff
@@ -10,6 +10,22 @@ $ yarn add -D sapui5-runtime
+
+where `X.Y.Z` is a valid version from [https://tools.hana.ondemand.com/](https://tools.hana.ondemand.com/#sapui5)
+
+Afterwards proceed with the normal installation 
```
Maybe we should change this line to: "Remember to run `npm install` or `yarn install` again after specifing your desired version."

This makes it a little more clear, that you need to run these command again, even if the runtime is already installed.

bastilimbach commented on README.md:
````diff
@@ -10,6 +10,22 @@ $ yarn add -D sapui5-runtime
+Add the following line to your `package.json`
+
+```JSON
+"sapui5RuntimeVersion": "X.Y.Z"
````
We should use a namespace for our settings to be future proof. I would suggest something like this:
```suggestion
"sapui5-runtime": {
  "version": "X.Y.Z"
}
```

With this change we could add more settings in the future, like specifying a custom download URL etc.

bastilimbach commented on src/install.js:
```diff
@@ -85,10 +85,21 @@ const versionEndpoint = url.parse('https://sapui5.hana.ondemand.com/resources/sa
+  let latestVersionURL
+  // Check if a specific Version of UI5 is needed..
+  // we are in node_modules/sapui5-runtime/src and need to go down to project using our module
+  const packageJson = fs.readJsonSync(path.join(__dirname, '../../../package.json'), { throws: false })
```
You can just require the `package.json`. I think this is a bit simpler than your approach. If you place this at the top of the file we can simply get a reference to the settings by using `config.setting`:
```javascript
const config = require(path.join(__dirname, '../../../package.json'))['sapui5-runtime'] || {}
```

As an example, to get the specified version we can use `config.version` and check if this is `undefined`.

mathiasmaerker replied:
Originally I used require but your eslint config forbid it and the travis build was failing [see here](https://travis-ci.org/bastilimbach/sapui5-runtime/builds/460775807) , so I changed it to fs-extra. I can change the .eslintrc to allow `global-require` but then I need to change `no-dynamic-require` as well. If that is ok for you I will do it, I just tried to adhere to your project guidelines :smile:

bastilimbach commented on src/install.js:
```diff
@@ -85,10 +85,21 @@ const versionEndpoint = url.parse('https://sapui5.hana.ondemand.com/resources/sa
-    const latestVersionURL = await determineLatestVersionURL(versionEndpoint, downloadEndpoint)
+    if (!latestVersionURL) {
+      latestVersionURL = await determineLatestVersionURL(versionEndpoint, downloadEndpoint)
+    }
```
With my suggested changes, this could be simplified to:
```javascript
if (!config.version) {
  latestVersionURL = await determineLatestVersionURL(versionEndpoint, downloadEndpoint)
} else {
  latestVersionURL = url.resolve(downloadEndpoint.href, `sapui5-rt-${config.version}.zip`)
}
```

bastilimbach reviewed (changes requested):
I just looked over your code and really like it. Could you please take a look at my purposed changes and implement them? I'm happy to merge your PR as soon as these are done.

At a later stage the package should also check which version of SAPUI5 is currently downloaded, so that we do not download the runtime on every `npm install`.

mathiasmaerker committed:
fix(#1): Implemented requested changes

+ disbaled elint
+ globle-imports
+ import/no-dynamic-require

+ changed property to object

bastilimbach commented:
I just manually merged your changes. Thanks again for your contribution! 👍
